presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java

```diff
@@ -428,7 +429,7 @@ public Node visitCreateFunction(SqlBaseParser.CreateFunctionContext context)
                 getType(context.returnType),
                 comment,
                 getRoutineCharacteristics(context.routineCharacteristics()),
-                (Return) visit(context.routineBody()));
+                (RoutineBody) visit(context.routineBody()));
```

Comment:
It only says that it's called an external routine, but I don't think it defines how external routine needs to be handled. I also didn't find anything specific about whether external routine requires a `<external body reference>`. So I'm not sure what to check here. Also the `<language name>` defined in SQL spec is very old. It doesn't have the languages we definitely want to support in the future like 'JAVA', 'PYTHON', 'PHP' or 'JAVASCRIPT' etc. Some of these could be defined inline potentially, so I don't want to rule that out.

Reply:
Those 2 lines of spec basically means:
- `LANGUAGE SQL` <=> SQL Routine <=> must have `SQL routine spec`, cannot have external routine body.
- `LANGUAGE not SQL` <=> external routine <=> must have external routine body, cannot have `SQL routine spec`.

Does those 2 checks make sense to you?

Reply:
That's not my understanding of the spec. And that's definitely not how other query engines that supports JS or LUA did it (they are not implemented as external functions). My understanding is literally, if the language is not SQL, is **called** *external routine* (which is a name that I didn't find any definition). I didn't find anything specify the relation ship between the name "external routine" and the spec `<external body reference>`. Would be great if you can point me to any documentation that made that association.

Reply:
11.60 Syntax Rules (1) pg. 1006:
```
An <SQL-invoked routine> specifies an SQL-invoked routine. Let R be the SQL-invoked routine specified by <SQL-invoked routine>.
```

11.60 Syntax Rule (23)(a) pg. 1015
```
If R is an SQL routine, then:
a) <SQL routine spec> shall be specified.
```

11.60 Syntax Rule (24)(a) pg. 1015
```
24) If R is an external routine, then:
a) <SQL routine spec> shall not be specified.
```
![image](https://user-images.githubusercontent.com/5472668/76564675-4b3dbb00-6466-11ea-8a9d-c9fcdd75ed50.png)
![image](https://user-images.githubusercontent.com/5472668/76564680-4e38ab80-6466-11ea-9c7c-11c5bebc8a67.png)

Reply:
If you are specifically saying that if the language is not SQL there should be no `RETURN xxx` I think that makes sense (at least for now). I don't think anything that's not SQL should have a `<external body reference>` though, even though that's also true for now. Since SQL spec does not define languages like JS or PYTHON, we need to extend the spec to support this. I don't think they have to have an `<external body reference>`, though they could. We can choose to not go against SQL spec (not using `RETURN definition`) so that language SQL <=> `RETURN expression` holds.